compiler/rustc_borrowck/src/lib.rs

```diff
@@ -499,7 +497,7 @@ impl<'tcx> BorrowckInferCtxt<'tcx> {
                 )
             });
 
-            self.inject_new_hidden_type_unchecked(key, hidden_ty);
+            assert_eq!(self.register_hidden_type_in_storage(key, hidden_ty), None);
```

Comment:
nit: code with side-effects in `assert!` macros kinda irks me, b/c it could easily accidentally be swapped to an `debug_assert_eq!` which affects whether this will run. could you split this into two statements?

Reply:
alternatively, why did you pull this assertion out of `inject_new_hidden_type_unchecked`?

Reply:
Will add a call-site which expects `Some` in #138785. Using `TypingMode::Borrowck` unfortunately changing currently non-defining uses to be defining, so I can't enable it with the old solver until I am done with the rest of the work. Having both behaviors in parallel is somewhat hard to support so I started by moving out some trivial changes out of that PR.